usr/man/man/usr/share/man/man1/ckpath.1
'\" te
.\"  Copyright 1989 AT&T  Copyright (c) 2005, Sun Microsystems, Inc.  All Rights Reserved
.TH ckpath 1 "4 Nov 2005" "SunOS 5.11" "User Commands"
.SH NAME
ckpath, errpath, helppath, valpath \- display a prompt; verify and return a pathname
.SH SYNOPSIS
.LP
.nf
\fBckpath\fR [\fB-Q\fR] [\fB-W\fR \fIwidth\fR] [\fB-a\fR | l] [\fB-b\fR | c | f | y] 
     [\fB-n\fR [o | z]] [\fB-rtwx\fR] [\fB-d\fR \fIdefault\fR] [\fB-h\fR \fIhelp\fR] 
     [\fB-e\fR \fIerror\fR] [\fB-p\fR \fIprompt\fR] [\fB-k\fR \fIpid\fR [\fB-s\fR \fIsignal\fR]]
.fi

.LP
.nf
\fB/usr/sadm/bin/errpath\fR [\fB-W\fR \fIwidth\fR] [\fB-a\fR | l] [\fB-b\fR | c | f | y] 
     [\fB-n\fR [o | z]] [\fB-rtwx\fR] [\fB-e\fR \fIerror\fR]
.fi

.LP
.nf
\fB/usr/sadm/bin/helppath\fR [\fB-W\fR \fIwidth\fR] [\fB-a\fR | l] [\fB-b\fR | c | f | y] 
     [\fB-n\fR [o | z]] [\fB-rtwx\fR] [\fB-h\fR \fIhelp\fR]
.fi

.LP
.nf
\fB/usr/sadm/bin/valpath\fR [\fB-a\fR | l] [\fB-b\fR | c | f | y] 
     [\fB-n\fR [o | z]] [\fB-rtwx\fR] \fIinput\fR
.fi

.SH DESCRIPTION
.sp
.LP
The \fBckpath\fR utility prompts a user and validates the response. It defines, among other things, a prompt message whose response should be a pathname, text for help and error messages, and a default value (which is returned if the user responds with a RETURN).
.sp
.LP
The pathname must obey the criteria specified by the first group of options. If no criteria is defined, the pathname must be for a normal file that does not yet exist. If neither \fB-a\fR (absolute) or \fB-l\fR (relative) is given, then either is assumed to be valid.
.sp
.LP
All messages are limited in length to 79 characters and are formatted automatically. Tabs and newlines are removed after a single white space character in a message definition, but spaces are not removed. When a tilde is placed at the beginning or end of a message definition, the default text is inserted at that point, allowing both custom text and the default text to be displayed.
.sp
.LP
If the prompt, help or error message is not defined, the default message (as defined under EXAMPLES) is displayed.
.sp
.LP
Three visual tool modules are linked to the \fBckpath\fR command. They are \fBerrpath\fR (which formats and displays an error message on the standard output), \fBhelppath\fR (which formats and displays a help message on the standard output), and \fBvalpath\fR (which validates a response). 
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.mk
.na
\fB\fB-a\fR\fR
.ad
.RS 14n
.rt  
Pathname must be an absolute path.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-b\fR\fR
.ad
.RS 14n
.rt  
Pathname must be a block special file.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-c\fR\fR
.ad
.RS 14n
.rt  
Pathname must be a character special file.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-d\fR \fIdefault\fR\fR
.ad
.RS 14n
.rt  
Defines the default value as \fIdefault\fR. The default is not validated and so does not have to meet any criteria.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-e\fR \fIerror\fR\fR
.ad
.RS 14n
.rt  
Defines the error message as \fI error\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-f\fR\fR
.ad
.RS 14n
.rt  
Pathname must be a regular file.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-h\fR \fIhelp\fR\fR
.ad
.RS 14n
.rt  
Defines the help message as \fI help\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-k\fR \fIpid\fR\fR
.ad
.RS 14n
.rt  
Specifies that process \fBID\fR \fIpid\fR is to be sent a signal if the user chooses to quit.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-l\fR\fR
.ad
.RS 14n
.rt  
Pathname must be a relative path.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-n\fR\fR
.ad
.RS 14n
.rt  
Pathname must not exist (must be new).
.RE

.sp
.ne 2
.mk
.na
\fB\fB-o\fR\fR
.ad
.RS 14n
.rt  
Pathname must exist (must be old).
.RE

.sp
.ne 2
.mk
.na
\fB\fB-p\fR \fIprompt\fR\fR
.ad
.RS 14n
.rt  
Defines the prompt message as \fIprompt\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-Q\fR\fR
.ad
.RS 14n
.rt  
Specifies that \fBquit\fR is not allowed as a valid response.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-r\fR\fR
.ad
.RS 14n
.rt  
Pathname must be readable.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-s\fR \fIsignal\fR\fR
.ad
.RS 14n
.rt  
Specifies that the process \fBID\fR \fIpid\fR defined with the \fB-k\fR option is to be sent signal \fIsignal\fR when quit is chosen. If no signal is specified, \fBSIGTERM\fR is used.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-t\fR\fR
.ad
.RS 14n
.rt  
Pathname must be creatable (touchable). Pathname will be created if it does not already exist.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-w\fR\fR
.ad
.RS 14n
.rt  
Pathname must be writable.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-W\fR \fIwidth\fR\fR
.ad
.RS 14n
.rt  
Specify that prompt, help and error messages be formatted to a line length of \fIwidth\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-x\fR\fR
.ad
.RS 14n
.rt  
Pathname must be executable.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-y\fR\fR
.ad
.RS 14n
.rt  
Pathname must be a directory.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-z\fR\fR
.ad
.RS 14n
.rt  
Pathname must have a file having a size greater than zero bytes.
.RE

.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.mk
.na
\fB\fIinput\fR\fR
.ad
.RS 9n
.rt  
Input to be verified against validation options.
.RE

.SH EXAMPLES
.sp
.LP
The text of the default messages for \fBckpath\fR depends upon  the criteria options that have been used.
.LP
\fBExample 1 \fRDefault prompt
.sp
.LP
An example default prompt for \fBckpath\fR (using the \fB-a\fR option) is:

.sp
.in +2
.nf
example% \fBckpath \fR\fB-a\fR
Enter an absolute pathname [?,q]
.fi
.in -2
.sp

.LP
\fBExample 2 \fRDefault error message
.sp
.LP
An example default error message (using the \fB-a\fR option) is:

.sp
.in +2
.nf
example% \fB/usr/sadm/bin/errpath \fR\fB-a\fR
ERROR: A pathname is a filename, optionally preceded by parent 
       directories.
       The pathname you enter: - must begin with a slash (/)
.fi
.in -2
.sp

.LP
\fBExample 3 \fRDefault help message
.sp
.LP
An example default help message (using the \fB-a\fR option) is:

.sp
.in +2
.nf
example% \fB/usr/sadm/bin/helppath \fR\fB-a\fR
A pathname is a filename, optionally preceded by parent directories.
The pathname you enter: - must begin with a slash (/)
.fi
.in -2
.sp

.LP
\fBExample 4 \fRThe quit option
.sp
.LP
When the quit option is chosen (and allowed), \fBq\fR is returned along with the return code \fB3\fR. Quit input gets a trailing newline.

.LP
\fBExample 5 \fRUsing the valpath module
.sp
.LP
The \fBvalpath\fR module will produce a usage message on stderr. It returns  \fB0\fR for success and non-zero for failure.

.sp
.in +2
.nf
example% \fB/usr/sadm/bin/valpath\fR
usage: valpath [\fB-[a|l][b|c|f|y][n|[o|z]]rtwx\fR] input
    .
    .
    .
.fi
.in -2
.sp

.SH EXIT STATUS
.sp
.LP
The following exit values are returned:
.sp
.ne 2
.mk
.na
\fB\fB0\fR\fR
.ad
.RS 5n
.rt  
Successful execution.
.RE

.sp
.ne 2
.mk
.na
\fB\fB1\fR\fR
.ad
.RS 5n
.rt  
\fBEOF\fR on input, or negative width on \fB-W\fR option, or usage error.
.RE

.sp
.ne 2
.mk
.na
\fB\fB2\fR\fR
.ad
.RS 5n
.rt  
Mutually exclusive options.
.RE

.sp
.ne 2
.mk
.na
\fB\fB3\fR\fR
.ad
.RS 5n
.rt  
User termination (quit).
.RE

.sp
.ne 2
.mk
.na
\fB\fB4\fR\fR
.ad
.RS 5n
.rt  
Mutually exclusive options.
.RE

.SH ATTRIBUTES
.sp
.LP
See \fBattributes\fR(5) for descriptions of the following attributes:
.sp

.sp
.TS
tab() box;
cw(2.75i) |cw(2.75i) 
lw(2.75i) |lw(2.75i) 
.
ATTRIBUTE TYPEATTRIBUTE VALUE
_
Availabilitysystem/core-os
.TE

.SH SEE ALSO
.sp
.LP
\fBsignal.h\fR(3HEAD), \fBattributes\fR(5)


usr/src/lib/libadm/common/ckpath.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/*
 * Copyright (c) 1989, 2011, Oracle and/or its affiliates. All rights reserved.
 */

#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "valtools.h"
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include "libadm.h"

#define	E_SYNTAX	"does not meet suggested filename syntax standard"
#define	E_READ		"is not readable"
#define	E_WRITE		"is not writable"
#define	E_EXEC		"is not executable"
#define	E_CREAT		"cannot be created"
#define	E_ABSOLUTE	"must begin with a slash (/)"
#define	E_RELATIVE	"must not begin with a slash (/)"
#define	E_EXIST		"does not exist"
#define	E_NEXIST	"must not already exist"
#define	E_BLK		"must specify a block special device"
#define	E_CHR		"must specify a character special device"
#define	E_DIR		"must specify a directory"
#define	E_REG		"must be a regular file"
#define	E_NONZERO	"must be a file of non-zero length"

#define	H_READ		"must be readable"
#define	H_WRITE		"must be writable"
#define	H_EXEC		"must be executable"
#define	H_CREAT		"will be created if it does not exist"
#define	H_ABSOLUTE	E_ABSOLUTE
#define	H_RELATIVE	E_RELATIVE
#define	H_EXIST		"must already exist"
#define	H_NEXIST	"must not already exist"
#define	H_BLK		E_BLK
#define	H_CHR		E_CHR
#define	H_DIR		E_DIR
#define	H_REG		E_REG
#define	H_NONZERO	E_NONZERO

#define	MSGSIZ	1024
#define	STDHELP \
	"A pathname is a filename, optionally preceded by parent directories."

static char	*errstr;
static char	*badset = "*?[]{}()<> \t'`\"\\|^";

static void
addhlp(char *msg, char *text)
{
	static int count;

	if (text == NULL) {
		count = 0;
		return;
	}
	if (!count++)
		(void) strcat(msg, " The pathname you enter:");
	(void) strcat(msg, "\\n\\t-\\ ");
	(void) strcat(msg, text);
}

static char *
sethlp(int pflags)
{
	char	*msg;

	msg = calloc(MSGSIZ, sizeof (char));
	addhlp(msg, NULL); /* initialize count */
	(void) strcpy(msg, STDHELP);

	if (pflags & P_EXIST)
		addhlp(msg, H_EXIST);
	else if (pflags & P_NEXIST)
		addhlp(msg, H_NEXIST);

	if (pflags & P_ABSOLUTE)
		addhlp(msg, H_ABSOLUTE);
	else if (pflags & P_RELATIVE)
		addhlp(msg, H_RELATIVE);

	if (pflags & P_READ)
		addhlp(msg, H_READ);
	if (pflags & P_WRITE)
		addhlp(msg, H_WRITE);
	if (pflags & P_EXEC)
		addhlp(msg, H_EXEC);
	if (pflags & P_CREAT)
		addhlp(msg, H_CREAT);

	if (pflags & P_BLK)
		addhlp(msg, H_BLK);
	else if (pflags & P_CHR)
		addhlp(msg, H_CHR);
	else if (pflags & P_DIR)
		addhlp(msg, H_DIR);
	else if (pflags & P_REG)
		addhlp(msg, H_REG);

	if (pflags & P_NONZERO)
		addhlp(msg, H_NONZERO);

	return (msg);
}

int
ckpath_stx(int pflags)
{
	if (((pflags & P_ABSOLUTE) && (pflags & P_RELATIVE)) ||
	    ((pflags & P_NEXIST) && (pflags &
	    (P_EXIST|P_NONZERO|P_READ|P_WRITE|P_EXEC))) ||
	    ((pflags & P_CREAT) && (pflags & (P_EXIST|P_NEXIST|P_BLK|P_CHR))) ||
	    ((pflags & P_BLK) && (pflags & (P_CHR|P_REG|P_DIR|P_NONZERO))) ||
	    ((pflags & P_CHR) && (pflags & (P_REG|P_DIR|P_NONZERO))) ||
	    ((pflags & P_DIR) && (pflags & P_REG))) {
		return (1);
	}
	return (0);
}

int
ckpath_val(char *path, int pflags)
{
	struct stat64 status;
	int	fd;
	char	*pt;

	if ((pflags & P_RELATIVE) && (*path == '/')) {
		errstr = E_RELATIVE;
		return (1);
	}
	if ((pflags & P_ABSOLUTE) && (*path != '/')) {
		errstr = E_ABSOLUTE;
		return (1);
	}
	if (stat64(path, &status)) {
		if (pflags & P_EXIST) {
			errstr = E_EXIST;
			return (1);
		}
		for (pt = path; *pt; pt++) {
			if (!isprint((unsigned char)*pt) ||
			    strchr(badset, *pt)) {
				errstr = E_SYNTAX;
				return (1);
			}
		}
		if (pflags & P_CREAT) {
			if (pflags & P_DIR) {
				if ((mkdir(path, 0755)) != 0) {
					errstr = E_CREAT;
					return (1);
				}
			} else {
				if ((fd = creat(path, 0644)) < 0) {
					errstr = E_CREAT;
					return (1);
				}
				(void) close(fd);
			}
		}
		return (0);
	} else if (pflags & P_NEXIST) {
		errstr = E_NEXIST;
		return (1);
	}
	if ((status.st_mode & S_IFMT) == S_IFREG) {
		/* check non zero status */
		if ((pflags & P_NONZERO) && (status.st_size < 1)) {
			errstr = E_NONZERO;
			return (1);
		}
	}
	if ((pflags & P_CHR) && ((status.st_mode & S_IFMT) != S_IFCHR)) {
		errstr = E_CHR;
		return (1);
	}
	if ((pflags & P_BLK) && ((status.st_mode & S_IFMT) != S_IFBLK)) {
		errstr = E_BLK;
		return (1);
	}
	if ((pflags & P_DIR) && ((status.st_mode & S_IFMT) != S_IFDIR)) {
		errstr = E_DIR;
		return (1);
	}
	if ((pflags & P_REG) && ((status.st_mode & S_IFMT) != S_IFREG)) {
		errstr = E_REG;
		return (1);
	}
	if ((pflags & P_READ) && !(status.st_mode & S_IREAD)) {
		errstr = E_READ;
		return (1);
	}
	if ((pflags & P_WRITE) && !(status.st_mode & S_IWRITE)) {
		errstr = E_WRITE;
		return (1);
	}
	if ((pflags & P_EXEC) && !(status.st_mode & S_IEXEC)) {
		errstr = E_EXEC;
		return (1);
	}
	return (0);
}

void
ckpath_err(int pflags, char *error, char *input)
{
	char	buffer[2048];
	char	*defhlp;

	if (input) {
		if (ckpath_val(input, pflags)) {
			(void) snprintf(buffer, sizeof (buffer),
			    "Pathname %s.", errstr);
			puterror(stdout, buffer, error);
			return;
		}
	}
	defhlp = sethlp(pflags);
	puterror(stdout, defhlp, error);
	free(defhlp);
}

void
ckpath_hlp(int pflags, char *help)
{
	char	*defhlp;

	defhlp = sethlp(pflags);
	puthelp(stdout, defhlp, help);
	free(defhlp);
}

int
ckpath(char *pathval, int pflags, char *defstr, char *error, char *help,
	char *prompt)
{
	char *defhlp;
	char input[MAX_INPUT];
	char buffer[256];

	if ((pathval == NULL) || ckpath_stx(pflags))
		return (2); /* usage error */

	if (!prompt) {
		if (pflags & P_ABSOLUTE)
			prompt = "Enter an absolute pathname";
		else if (pflags & P_RELATIVE)
			prompt = "Enter a relative pathname";
		else
			prompt = "Enter a pathname";
	}
	defhlp = sethlp(pflags);

start:
	putprmpt(stderr, prompt, NULL, defstr);
	if (getinput(input)) {
		free(defhlp);
		return (1);
	}

	if (strlen(input) == 0) {
		if (defstr) {
			(void) strcpy(pathval, defstr);
			free(defhlp);
			return (0);
		}
		puterror(stderr, NULL, "Input is required.");
		goto start;
	}
	if (strcmp(input, "?") == 0) {
		puthelp(stderr, defhlp, help);
		goto start;
	}
	if (ckquit && (strcmp(input, "q") == 0)) {
		free(defhlp);
		return (3);
	}

	if (ckpath_val(input, pflags)) {
		(void) snprintf(buffer, sizeof (buffer), "Pathname %s.",
		    errstr);
		puterror(stderr, buffer, error);
		goto start;
	}
	(void) strcpy(pathval, input);
	free(defhlp);
	return (0);
}
